File: go/leetcode.go
package main

import (
	"fmt"
	"math"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func leetcode2(l1 *ListNode, l2 *ListNode) *ListNode {
	var head, tail *ListNode
	carry := 0
	for l1 != nil || l2 != nil {
		n1, n2 := 0, 0
		if l1 != nil {
			n1 = l1.Val
		}
		if l2 != nil {
			n2 = l2.Val
		}
		sum := n1 + n2 + carry
		if head == nil {
			head = &ListNode{Val: sum % 10}
			tail = head
		} else {
			tail.Next = &ListNode{Val: sum % 10}
			tail = tail.Next
		}
		carry = sum / 10
		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}
	}
	if carry > 0 {
		tail.Next = &ListNode{
			Val: carry,
		}
	}
	return head
}

func leetcode26(nums []int) int {
	slow, fast := 0, 1
	for fast < len(nums) {
		if nums[slow] != nums[fast] {
			slow++
			nums[slow] = nums[fast]
		}
		fast++
	}
	return slow + 1
}

func leetcode27(nums []int, val int) int {
	var index, resultSzie int
	for _, i := range nums {
		if i != val {
			nums[index] = i
			index++
			resultSzie++
		}
	}
	return resultSzie
}

func leetcode80(nums []int) int {
	if len(nums) <= 2 {
		return len(nums)
	}

	slow, fast := 2, 2
	for fast < len(nums) {
		if nums[fast] != nums[slow-2] {
			nums[slow] = nums[fast]
			slow++
		}
		fast++
	}
	return slow
}

func leetcode88(nums1 []int, m int, nums2 []int, n int) {
	s := m + n - 1
	for m > 0 && n > 0 {
		if nums1[m-1] >= nums2[n-1] {
			nums1[s] = nums1[m-1]
			s--
			m--
		} else {
			nums1[s] = nums2[n-1]
			s--
			n--
		}
	}
	for i := n; i > 0; i-- {
		nums1[s] = nums2[i-1]
		s--
	}
}

func leetcode121(prices []int) int {
	minPrice, maxProfit := math.MaxInt, 0
	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

func leetcode128(nums []int) int {
	numSet := map[int]bool{}
	for _, num := range nums {
		numSet[num] = true
	}
	result := 0
	for i := range numSet {
		if !numSet[i-1] {
			cur := i
			curLength := 1
			for numSet[cur+1] {
				cur++
				curLength++
			}
			if curLength > result {
				result = curLength
			}
		}
	}
	return result
}

type leetcode155 struct {
	stack    []int
	minStack []int
}

func Constructor() leetcode155 {
	return leetcode155{
		stack:    []int{},
		minStack: []int{math.MaxInt64},
	}
}

func (this *leetcode155) Push(val int) {
	this.stack = append(this.stack, val)
	this.minStack = append(this.minStack, min(val, this.minStack[len(this.minStack)-1]))
}

func (this *leetcode155) Pop() {
	this.stack = this.stack[:len(this.stack)-1]
	this.minStack = this.minStack[:len(this.minStack)-1]
}

func (this *leetcode155) Top() int {
	return this.stack[len(this.stack)-1]
}

func (this *leetcode155) GetMin() int {
	return this.minStack[len(this.minStack)-1]
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func leetcode169(nums []int) int {
	count, result := 0, 0
	for _, v := range nums {
		if count == 0 {
			result = v
		}
		if v == result {
			count++
		} else {
			count--
		}
	}
	return result
}

func leetcode189(nums []int, k int) {
	k %= len(nums)
	if k == 0 || len(nums) == 1 {
		return
	}
	reverseArray(nums, 0, len(nums)-1)
	reverseArray(nums, 0, k-1)
	reverseArray(nums, k, len(nums)-1)
}

func reverseArray(nums []int, i int, j int) {
	for i < j {
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--
	}
}

func main() {
	fmt.Println("hello world")
}

File: python3/leetcode.py
import math
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def addTwoNumbers(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    head = None
    tail = None
    carry = 0
    while l1 is not None or l2 is not None:
        n1 = n2 = 0
        if l1 is not None:
            n1 = l1.val
        if l2 is not None:
            n2 = l2.val
        sum_val = n1 + n2 + carry
        carry = int(sum_val / 10)
        if head is None:
            head = tail = ListNode(val=sum_val % 10)
        else:
            tail.next = ListNode(val=sum_val % 10)
            tail = tail.next
        if l1 is not None:
            l1 = l1.next
        if l2 is not None:
            l2 = l2.next
    if carry > 0:
        tail.next = ListNode(val=carry)
    return head


def leetcode26(self, nums: List[int]) -> int:
    slow, fast = 0, 1
    while fast < len(nums):
        if nums[fast] != nums[slow]:
            slow += 1
            nums[slow] = nums[fast]
        fast += 1
    return slow + 1


def leetcode27(nums: List[int], val: int) -> int:
    index = 0
    size = 0
    for v in nums:
        if v != val:
            nums[index] = v
            index = index + 1
            size = size + 1
    return size


def leetcode80(self, nums: List[int]) -> int:
    if len(nums) <= 2:
        return len(nums)
    slow, fast = 2, 2
    while fast < len(nums):
        if nums[fast] != nums[slow - 2]:
            nums[slow] = nums[fast]
            slow += 1
        fast += 1
    return slow


def leetcode88(nums1: List[int], m: int, nums2: List[int], n: int) -> None:
    """
    Do not return anything, modify nums1 in-place instead.
    """
    s = m + n - 1
    while m > 0 and n > 0:
        if nums1[m - 1] >= nums2[n - 1]:
            nums1[s] = nums1[m - 1]
            s -= 1
            m -= 1
        else:
            nums1[s] = nums2[n - 1]
            s -= 1
            n -= 1
    while n > 0:
        nums1[s] = nums2[n - 1]
        s -= 1
        n -= 1


def leetcode121(self, prices: List[int]) -> int:
    minPrice, maxProfit = prices[0], 0
    for price in prices:
        if price < minPrice:
            minPrice = price
        elif price - minPrice > maxProfit:
            maxProfit = price - minPrice
    return maxProfit


def leetcode128(self, nums: List[int]) -> int:
    result = 0
    num_set = set(nums)
    for e in num_set:
        if e - 1 not in num_set:
            cur = e
            curLength = 1
            while cur + 1 in num_set:
                cur += 1
                curLength += 1
            result = max(result, curLength)
    return result


class leetcode155:

    def __init__(self):
        self.x_stack = []
        self.min_stack = [math.inf]

    def push(self, val: int) -> None:
        self.x_stack.append(val)
        self.min_stack.append(min(val, self.min_stack[-1]))

    def pop(self) -> None:
        self.x_stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.x_stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


def leetcode169(self, nums: List[int]) -> int:
    count, result = 0, 0
    for v in nums:
        if count == 0:
            result = v
        if v == result:
            count += 1
        else:
            count -= 1
    return result


def leetcode189(self, nums: List[int], k: int) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    k %= len(nums)
    if k == 0 or len(nums) == 1:
        return
    i, j = 0, len(nums) - 1
    while i < j:
        nums[i], nums[j] = nums[j], nums[i]
        i += 1
        j -= 1

    i, j = 0, k - 1
    while i < j:
        nums[i], nums[j] = nums[j], nums[i]
        i += 1
        j -= 1

    i, j = k, len(nums) - 1
    while i < j:
        nums[i], nums[j] = nums[j], nums[i]
        i += 1
        j -= 1


if __name__ == '__main__':
    print("ok")
